src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -614,6 +614,77 @@ QString QgsPostgresUtils::whereClause( const QgsFeatureIds &featureIds, const Qg
     return expr;
   };
 
+  auto lookupComboKeyWhereClause = [ = ]( bool canUseVALUES, QList<QString> withCastTypeName )
+  {
+    if ( pkAttrs.size() == 1 )
+      return lookupKeyWhereClause();
+
+    if ( featureIds.isEmpty() )
+      return QString();
+
+    QString delim;
+    QString whereKeys;
+    for ( int i = 0; i < pkAttrs.size(); i++ )
+    {
+      whereKeys += delim + QgsPostgresConn::quotedIdentifier( fields.at( pkAttrs[i] ).name() );
+      delim = QStringLiteral( "," );
+    }
+
+    QgsFeatureIds fidsToOr;
+    QStringList whereValuesList;
+    for ( const QgsFeatureId featureId : std::as_const( featureIds ) )
+    {
+      const QVariantList pkVals = sharedData->lookupKey( featureId );
+      if ( !pkVals.isEmpty() )
+      {
+        Q_ASSERT( pkVals.size() == pkAttrs.size() );
+
+        QString delim;
+        QString whereValues;
+        bool isNotNull = true;
+        for ( int i = 0; i < pkAttrs.size(); i++ )
+        {
+          if ( pkVals[i].isNull() )
+            isNotNull = false;
+          whereValues += delim + QgsPostgresConn::quotedValue( pkVals[i] );
+          if ( canUseVALUES && withCastTypeName.contains( fields.at( pkAttrs[i] ).typeName() ) )
+            whereValues += QStringLiteral( "::%1" ).arg( fields.at( pkAttrs[i] ).typeName() );
+          delim = QStringLiteral( "," );
+        }
+        if ( isNotNull )
+          whereValuesList << whereValues;
+        else
+          fidsToOr.insert( featureId );
+      }
+    }
+
+    QString whereInString;
+    if ( canUseVALUES )
+      whereInString = whereValuesList.isEmpty() ? QString() :
+                      whereKeys.prepend( QLatin1Char( '(' ) ).append( QStringLiteral( ") IN " ) ) +
+                      whereValuesList.join( QStringLiteral( "),(" ) ).prepend( QStringLiteral( "( VALUES (" ) ).append( QStringLiteral( ") )" ) );
+    else
+      whereInString = whereValuesList.isEmpty() ? QString() :
+                      whereKeys.prepend( QLatin1Char( '(' ) ).append( QStringLiteral( ") IN " ) ) +
+                      whereValuesList.join( QStringLiteral( "),(" ) ).prepend( QStringLiteral( "( (" ) ).append( QStringLiteral( ") )" ) );
```

Comment:
```suggestion
      whereInString = whereValuesList.isEmpty() ? QString() :
                      whereKeys.prepend( QLatin1Char( '(' ) ).append( QStringLiteral( ") IN " ) ) +
                      whereValuesList.join( QStringLiteral( "),(" ) ).prepend( QStringLiteral( "( %1 (" ) ).append( QStringLiteral( ") )" ) );
                      whereInString.arg( canUseValues ? "VALUES" : QString() );
```
A more concise way maybe